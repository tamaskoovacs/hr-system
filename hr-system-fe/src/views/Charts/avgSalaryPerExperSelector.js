import tableData from "static/tableData.json";

const calcAvgSalaryPerExper = () => {
  const datasets = {};

  tableData?.forEach((row) => {
    const exp = row?.years_of_experience;

    if (exp) {
      datasets[exp] = {
        count: datasets[exp]?.count ? datasets[exp]?.count + 1 : 1,
        totalSalary: datasets[exp]?.totalSalary ? datasets[exp]?.totalSalary + row?.salary : row?.salary,
      };
    }
  });

  const result = { labels: [], data: [] };
  Object?.entries(datasets)?.forEach(([key, value]) => {
    result?.labels?.push(key);
    result?.data?.push(value?.totalSalary / value?.count);
  });

  return result;
};

export default calcAvgSalaryPerExper;
