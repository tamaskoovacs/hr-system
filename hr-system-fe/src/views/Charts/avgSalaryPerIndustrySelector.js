import tableData from "static/tableData.json";

const calcAvgSalaryPerIndustry = () => {
  const datasets = {};

  tableData?.forEach((row) => {
    const industry = row?.industry;
    if (industry) {
      datasets[industry] = {
        count: datasets[industry]?.count ? datasets[industry]?.count + 1 : 1,
        salarySum: datasets[industry]?.salarySum ? datasets[industry]?.salarySum + row?.salary : row?.salary,
      };
    }
  });

  const result = { labels: [], data: [] };
  Object?.entries(datasets)?.forEach(([key, value]) => {
    result?.labels?.push(key);
    result?.data?.push(value?.salarySum / value?.count);
  });

  return result;
};

export default calcAvgSalaryPerIndustry;
