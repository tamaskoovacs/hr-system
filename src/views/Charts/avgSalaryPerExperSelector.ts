import { datasets, result } from 'interfaces/interfaces';
import tableData from 'static/tableData.json';

const calcAvgSalaryPerExper = () => {
  const datasets: datasets = {};

  tableData?.forEach((row) => {
    const exp = row?.years_of_experience;

    if (exp) {
      datasets[exp] = {
        count: datasets[exp]?.count ? datasets[exp]?.count + 1 : 1,
        total: datasets[exp]?.total ? datasets[exp]?.total + row?.salary : row?.salary
      };
    }
  });

  const result: result = { labels: [], data: [] };
  Object?.entries(datasets)?.forEach(([key, value]) => {
    result?.labels?.push(key);
    result?.data?.push(value?.total / value?.count);
  });

  return result;
};

export default calcAvgSalaryPerExper;
