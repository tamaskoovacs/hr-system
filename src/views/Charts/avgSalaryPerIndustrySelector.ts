import { datasets, result } from 'interfaces/interfaces';
import tableData from 'static/tableData.json';

const calcAvgSalaryPerIndustry = () => {
  const datasets: datasets = {};

  tableData?.forEach((row) => {
    const industry = row?.industry;
    if (industry) {
      datasets[industry] = {
        count: datasets[industry]?.count ? datasets[industry]?.count + 1 : 1,
        total: datasets[industry]?.total ? datasets[industry]?.total + row?.salary : row?.salary
      };
    }
  });

  const result: result = { labels: [], data: [] };
  Object?.entries(datasets)?.forEach(([key, value]) => {
    result?.labels?.push(key);
    result?.data?.push(value?.total / value?.count);
  });

  return result;
};

export default calcAvgSalaryPerIndustry;
