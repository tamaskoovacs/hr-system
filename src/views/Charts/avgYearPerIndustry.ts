import tableData from 'static/tableData.json';
import moment from 'moment';
import { datasets, result } from 'interfaces/interfaces';

const avgYearPerIndustry = () => {
  const datasets: datasets = {};

  tableData?.forEach((row) => {
    const date = row?.date_of_birth?.split('/');
    const newDate = `${date[1]}/${date[0]}/${date[2]}`;
    const diff = moment(new Date()).diff(moment(new Date(newDate)), 'years');
    const industry = row?.industry;

    if (industry) {
      datasets[industry] = {
        count: datasets[industry]?.count ? datasets[industry]?.count + 1 : 1,
        total: datasets[industry]?.total ? datasets[industry]?.total + diff : diff
      };
    }
  });

  const result: result = { labels: [], data: [] };
  Object?.entries(datasets)?.forEach(([key, value]) => {
    result?.labels?.push(key);
    result?.data?.push(value?.total / value?.count);
  });

  return result;
};

export default avgYearPerIndustry;
