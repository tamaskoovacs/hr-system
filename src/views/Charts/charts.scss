.chart-container > canvas {
  margin: 50px;
}
