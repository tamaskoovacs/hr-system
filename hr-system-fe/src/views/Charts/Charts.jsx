import { Chart as ChartJS, CategoryScale, LinearScale, BarElement, Title, Tooltip } from "chart.js";
import { Bar } from "react-chartjs-2";
import calcAvgSalaryPerIndustry from "./avgSalaryPerIndustrySelector";
import './charts.scss'

ChartJS.register(CategoryScale, LinearScale, BarElement, Title, Tooltip);

const getOptions = (title) => {
    return {
        responsive: true,
        plugins: {
            legend: {
            display: false,
            },
            title: {
            display: true,
            text: title,
            },
        },
    }
}

const getData = (labels, data, color) => {
    return {
        labels,
        datasets: [
            {
                data: data,
                backgroundColor: color,
            }
        ]
    }
} 

const Charts = () => {
  const avgSalarPerIndustry = calcAvgSalaryPerIndustry();

  return (
    <div className="chart-container">
      <Bar options={getOptions('Avarage salary per industry')} data={getData(avgSalarPerIndustry?.labels, avgSalarPerIndustry?.data, 'red')} />
    </div>
  );
};

export default Charts;