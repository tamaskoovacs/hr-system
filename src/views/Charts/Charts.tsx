import { Chart as ChartJS, CategoryScale, LinearScale, BarElement, Title, Tooltip } from 'chart.js';
import { Bar } from 'react-chartjs-2';
import calcAvgSalaryPerIndustry from './avgSalaryPerIndustrySelector';
import calcAvgSalaryPerExper from './avgSalaryPerExperSelector';
import calcAvgYearPerIndustry from './avgYearPerIndustry';
import './charts.scss';
import { useNavigate } from 'react-router-dom';

ChartJS.register(CategoryScale, LinearScale, BarElement, Title, Tooltip);

const getOptions = (title: string) => {
  return {
    responsive: true,
    plugins: {
      legend: {
        display: false
      },
      title: {
        display: true,
        text: title
      }
    }
  };
};

const getData = (labels: string[], data: number[], color: string) => {
  return {
    labels,
    datasets: [
      {
        data: data,
        backgroundColor: color
      }
    ]
  };
};

const Charts = () => {
  let navigate = useNavigate();
  const handleChartsClick = () => {
    navigate('/');
  };

  const avgSalarPerIndustry = calcAvgSalaryPerIndustry();
  const avgSalaryPerExper = calcAvgSalaryPerExper();
  const avgYearPerIndustry = calcAvgYearPerIndustry();

  return (
    <div className="chart-container">
      <div className="top-section">
        <h1>Charts</h1>
        <button onClick={handleChartsClick}>Hr System Board</button>
      </div>

      <Bar
        options={getOptions('Avarage salary per industry')}
        data={getData(avgSalarPerIndustry?.labels, avgSalarPerIndustry?.data, 'red')}
      />
      <Bar
        options={getOptions('Avarage salary per year of experience')}
        data={getData(avgSalaryPerExper?.labels, avgSalaryPerExper?.data, 'green')}
      />
      <Bar
        options={getOptions('Avarage year per industry')}
        data={getData(avgYearPerIndustry?.labels, avgYearPerIndustry?.data, 'blue')}
      />
    </div>
  );
};

export default Charts;
