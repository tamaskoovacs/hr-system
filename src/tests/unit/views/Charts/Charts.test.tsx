import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Charts from 'views/Charts/Charts';

describe('Charts page', () => {
  test('It should matches snapshot', () => {
    const view = render(
      <BrowserRouter>
        <Charts />
      </BrowserRouter>
    );
    expect(view.container).toMatchSnapshot();
  });
});
