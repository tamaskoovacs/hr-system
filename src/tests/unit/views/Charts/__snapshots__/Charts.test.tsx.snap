// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts page It should matches snapshot 1`] = `
<div>
  <div
    class="chart-container"
  >
    <div
      class="top-section"
    >
      <h1>
        Charts
      </h1>
      <button>
        Hr System Board
      </button>
    </div>
    <canvas
      height="150"
      role="img"
      width="300"
    />
    <canvas
      height="150"
      role="img"
      width="300"
    />
    <canvas
      height="150"
      role="img"
      width="300"
    />
  </div>
</div>
`;
